fe2-android/app/src/test/java/com/github/dedis/student20_pop/model/data/LAORepositoryTest.java

```diff
@@ -0,0 +1,308 @@
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import junit.framework.TestCase;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class LAORepositoryTest extends TestCase {
+
+  @Mock
+  LAORemoteDataSource remoteDataSource;
+
+  @Mock
+  LAOLocalDataSource localDataSource;
+
+  @Mock
+  AndroidKeysetManager androidKeysetManager;
+
+  @Mock
+  PublicKeySign signer;
+
+  private static final int REQUEST_ID = 42;
+  private static final int RESPONSE_DELAY = 1000;
+  private static final String ORGANIZER = "Z3DYtBxooGs6KxOAqCWD3ihR8M6ZPBjAmWp_w5VBaws=";
+  private static final String LAO_NAME = "Lao";
+  private static final CreateLao CREATE_LAO = new CreateLao(LAO_NAME, ORGANIZER);
+  private static final String CHANNEL = "/root";
+  private static final String LAO_CHANNEL = CHANNEL + "/" + CREATE_LAO.getId();
+
+  private LAORepository repository;
+  private TestScheduler testScheduler;
+  private MessageGeneral createLaoMessage;
+  private TestObserver<Answer> answerTestObserver;
+  private TestObserver<Lao> laoTestObserver;
+  private TestObserver<List<Lao>> laoListTestObserver;
+  private SchedulerProvider testSchedulerProvider;
+
+  @Before
+  public void setup() throws GeneralSecurityException {
+    testSchedulerProvider = new TestSchedulerProvider();
+    testScheduler = (TestScheduler) testSchedulerProvider.io();
+
+    // Mock the signing of of the data CreateLao
+    byte[] dataBuf = Injection.provideGson().toJson(CREATE_LAO, Data.class).getBytes();
+    Mockito.when(signer.sign(dataBuf)).thenReturn(dataBuf);
+
+    // Create the message containing a CreateLao data
+    createLaoMessage = new MessageGeneral(Base64.getUrlDecoder().decode(ORGANIZER), CREATE_LAO, signer, Injection.provideGson());
+
+    // Check the correct data is signed during the creation of MessageGeneral
+    Mockito.verify(signer, Mockito.times(1)).sign(dataBuf);
+
+    // Mock the remote data source to always return the same request id
+    Mockito.when(remoteDataSource.incrementAndGetRequestId()).thenReturn(REQUEST_ID);
+    Mockito.when(remoteDataSource.observeWebsocket()).thenReturn(Observable.empty());
+
+    // Create the observer that will simulate the network answer
+    answerTestObserver = TestObserver.create();
+
+    // Create the observers for the Repository's LAO lists to test the LAO creation and subscription
+    laoTestObserver = TestObserver.create();
+    laoListTestObserver = TestObserver.create();
+  }
+
+  @After
+  public void destroy() {
+    // Ensure every test has a new LAORepository instance with a different TestSchedulerProvider
+    LAORepository.destroyInstance();
+  }
+
+  @Test
+  public void testSendCatchup() {
+    // Simulate a network response from the server after the response delay
```

Comment:
(I'm not experienced with this code base, so I might be very wrong) It feels like this is more testing the testing framework than the system itself. Here we are testing the sending of a message. The *only* thing we should test is if our system, given we sent a catchup message and we then got an answer (which we can assume since we are mocking it), did what we expected it to do (i.e use the Answer, process it, etc...). (And then we could also test if the system reacts correctly in the case the answer doesn't come back)